sys/vane/clay.hoon

```diff
@@ -2314,6 +2360,66 @@
         ?^(r.mun ~ !!) :: [~ %w !>([t.yak (forge-nori yak)])])-all
       (query(ank.dom ank:(descend-path:(zu ank.dom) r.mun)) p.mun) ::  dead code
     ::
+    ::  Gets the permissions that apply to a particular node.
+    ::
+    ::  If the node has no permissions of its own, we use its parent's.
+    ::  If no permissions have been set for the entire tree above the node,
+    ::  we default to fully private (empty whitelist).
+    ::
+    ++  read-p
+      |=  pax/path
+      ^-  (unit (unit (each cage lobe)))
+      =-  [~ ~ %& %noun !>(-)]
+      :-  (read-p-in pax per.red)
+      (read-p-in pax pew.red)
+    ::
+    ++  read-p-in
+      |=  {pax/path pes/(map path rule)}
+      ^-  dict
+      =+  rul=(~(get by pes) pax)
+      ?^  rul  [pax u.rul]
+      ?~  pax  [/ %white ~]
+      $(pax t.pax)
+    ::
+    ++  may-read
+      |=  {who/ship car/care yon/aeon pax/path}
+      ^-  ?
+      ?+  car
+        (allowed-by who pax per.red)
+      ::
+          $p
+        =(who our)
+      ::
+          ?($y $z)
+        =+  tak=(~(get by hit.dom) yon)
+        ?~  tak  |
+        =+  yak=(tako-to-yaki u.tak)
+        =+  len=(lent pax)
+        =-  (levy ~(tap in -) |=(p/path (allowed-by who p per.red)))
+        %+  roll  ~(tap in (~(del in ~(key by q.yak)) pax))
+        |=  {p/path s/(set path)}
+        ?.  =(pax (scag len p))  s
+        %-  ~(put in s)
+        ?:  ?=($z car)  p
+        (scag +(len) p)
+      ==
+    ::
+    ++  may-write
+      |=  {w/ship p/path}
+      (allowed-by w p pew.red)
+    ::
+    ++  allowed-by
+      |=  {who/ship pax/path pes/(map path rule)}
+      ^-  ?
+      =+  rul=rul:(read-p-in pax pes)
+      =-  =(0 -)
+      %+  mix  ?=($black mod.rul)
+      %-  ~(rep in who.rul)
+      |=  {w/whom h/?}
+      ?:  h  &
+      ?:  ?=($& -.w)  =(w &+who)
+      (~(has in (fall (~(get by cez) p.w) ~)) who)
```

Comment:
Could `(~(get by cez) p.w)` ever fail? I feel like that should be a `~(got by ...)`

Reply:
It totally can. Either I added a group which I haven't defined yet, or I deleted a group and there's still references to it lingering.